File: src/main/go/leetcode/lists/hot100/LeetCode31.go
package hot100

func nextPermutation(nums []int) {

	if len(nums) == 0 {
		return
	}

	i := len(nums) - 2
	for i >= 0 && nums[i] >= nums[i+1] {
		i--
	}
	if i >= 0 {
		j := len(nums) - 1
		for j >= 0 && nums[j] <= nums[i] {
			j--
		}
		nums[i], nums[j] = nums[j], nums[i]
	}

	Reverse(nums[i+1:])
}

File: java/src/leetcode/LeetCode31.java
package leetcode;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/16 23:08
 */
public class LeetCode31 {

    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length == 0) {
            return;
        }

        int i = nums.length - 1;
        while (i > 0 && nums[i] <= nums[i - 1]) {
            i--;
        }
        if (i == 0) {
            reverse(nums, 0, nums.length - 1);
            return;
        }
        int index = i - 1;
        int j = nums.length - 1;

        while (j > 0 && nums[j] <= nums[index]) {
            j--;
        }

        swap(nums, index, j);
        reverse(nums, index + 1, nums.length - 1);
    }

    private void swap(int[] nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }

    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            int t = nums[start];
            nums[start] = nums[end];
            nums[end] = t;
            start++;
            end--;
        }
    }
}
